### ISSUE TITLE
[mlir][mpi] MPIToLLVM multiplies memref element counts in i32

### ISSUE THREAD

**Issue by snarang181** [llvm/llvm-project#225257]
## Summary

MPIToLLVM truncates every memref extent to i32 and multiplies the shape in
i32 before calling the ordinary MPI C interface. A verifier-clean
`memref<65536x65536xi8>` has `2^32` elements, but the generated `MPI_Send`
count wraps to zero, silently turning a full send into an empty send.

## Reproducer

```mlir
module attributes {dlti.map=#dlti.map<"MPI:Implementation"="MPICH">} {
 memref.global @buf : memref<65536x65536xi8> = uninitialized
 func.func @case(%rank: i32) {
  %b = memref.get_global @buf : memref<65536x65536xi8>
  %c = mpi.comm_world : !mpi.comm
  mpi.send(%b,%rank,%rank,%c) : memref<65536x65536xi8>,i32,i32
  return
 }
}
```

Both `b2b38d4b3352` and `57b5017e0102` produce:

```console
$ mlir-opt repro.mlir --convert-to-llvm='dynamic=true' | grep -E 'constant\(65536|llvm.mul|llvm.call @MPI_Send'
    %0 = llvm.mlir.constant(65536 : i64) : i64
    %1 = llvm.mlir.constant(65536 : i64) : i64
    %27 = llvm.mul %26, %24 : i32
    %30 = llvm.mul %29, %27 : i32
    %33 = llvm.call @MPI_Send(%23, %30, %31, %arg0, %arg0, %32) : (!llvm.ptr, i32, i32, i32, i32, i32) -> i32
```

Expected count: `4294967296`, or a diagnostic/large-count MPI binding.
Actual signed-i32 count: `65536*65536 mod 2^32 = 0`.

## Agent Analysis

At `b2b38d4b3352`, `mlir/lib/Conversion/MPIToLLVM/MPIToLLVM.cpp:53-88`
creates an i32 accumulator, truncates descriptor extents, and multiplies in
i32. Compute with checked descriptor-width arithmetic and select `MPI_*_c`
when available, otherwise reject or guard unrepresentable counts.

**Related:** PR #178870 fixed rank>1 counting by multiplying every extent, but
retained the i32 accumulator and did not address count representability.
